File: Trumpfish.WebClient/src/features/biddingBrowser/conditionReader.ts
import { toNumber, type NumberRange } from '@/api/models';
import { rangeFields, type InheritedRange, type InheritedRanges, type RangeField } from './constraints';

/** Which field each suit marker fills. Suits are told apart by the first two letters of the word, as descriptions write them. */
const suitFields: Record<'tr' | 'ka' | 'ki' | 'pi', RangeField> = {
  tr: 'clubsCardRange',
  ka: 'diamondsCardRange',
  ki: 'heartsCardRange',
  pi: 'spadesCardRange',
};

type SuitMarker = keyof typeof suitFields;
type Bounds = { lower: number | null; upper: number | null };

/**
 * `ka` must not swallow "kart", "kartowy" or "karty": those count cards rather than name diamonds, and they sit in exactly the
 * position a suit would ("4+ kartowy fit" is a four card fit). All three start with "kart", so one lookahead rules them out.
 */
const suitMarkers = String.raw`tr|ka(?!rt)|ki|pi`;

/** Point ranges are written with a hyphen about as often as with a dash. */
const dash = String.raw`[-–—]`;

const pointsPatterns: { pattern: RegExp; read: (match: RegExpExecArray) => NumberRange }[] = [
  { pattern: new RegExp(String.raw`(\d+)\s*${dash}\s*(\d+)\s*pc\b`, 'i'), read: (match) => ({ lower: Number(match[1]), upper: Number(match[2]) }) },
  { pattern: /(\d+)\s*\+\s*pc\b/i, read: (match) => ({ lower: Number(match[1]), upper: null }) },
  // "Poniżej N PC" names the bound rather than excluding it: it means N or fewer, not N-1 or fewer.
  { pattern: /poni[żz]ej\s+(\d+)\s*pc\b/i, read: (match) => ({ lower: null, upper: Number(match[1]) }) },
];

const suitPatterns: { pattern: RegExp; read: (count: number) => Partial<Bounds> }[] = [
  { pattern: new RegExp(String.raw`(\d+)\s*\+\s*(${suitMarkers})`, 'gi'), read: (count) => ({ lower: count }) },
  { pattern: new RegExp(String.raw`brak\s+(\d+)\s*(${suitMarkers})`, 'gi'), read: (count) => ({ upper: Math.max(0, count - 1) }) },
  { pattern: new RegExp(String.raw`dok[łl]adnie\s+(\d+)\s*(${suitMarkers})`, 'gi'), read: (count) => ({ lower: count, upper: count }) },
];

/**
 * Reads the point range and the suit lengths out of a bid's description, as a patch holding only what could actually be read.
 * Everything else on the bid is left alone, so running this again overwrites the facts it recognises and nothing more.
 *
 * The description is taken a comma separated section at a time; a section that matches nothing is simply skipped, which is what
 * lets the prose that surrounds the numbers ("blokujące", "ręka zrównoważona") sit there untouched.
 *
 * A range that comes out exactly as the player already promised it is dropped: writing it again would state explicitly what the
 * sequence already implies, and the editor shows it as a placeholder either way.
 */
export function readCondition(text: string, inherited: InheritedRanges): Partial<Record<RangeField, NumberRange>> {
  const found: Partial<Record<RangeField, NumberRange>> = {};
  const suits = new Map<SuitMarker, Bounds>();

  for (const section of text.split(',')) {
    // Points are written once, at the front, so the first section that yields them wins.
    if (found.pointsRange === undefined) {
      const points = readPoints(section);
      if (points !== null) {
        found.pointsRange = points;
      }
    }

    readSuits(section, suits);
  }

  for (const [marker, bounds] of suits) {
    found[suitFields[marker]] = bounds;
  }

  const patch: Partial<Record<RangeField, NumberRange>> = {};
  for (const field of rangeFields) {
    const range = found[field];
    if (range !== undefined && !repeatsInherited(range, inherited[field])) {
      patch[field] = range;
    }
  }

  return patch;
}

/** Whether a range says precisely what the ancestors already said, in which case restating it on this bid adds nothing. */
function repeatsInherited(range: NumberRange, inherited: InheritedRange | undefined): boolean {
  return inherited !== undefined && toNumber(range.lower) === inherited.lower && toNumber(range.upper) === inherited.upper;
}

function readPoints(section: string): NumberRange | null {
  for (const { pattern, read } of pointsPatterns) {
    const match = pattern.exec(section);
    if (match !== null) {
      return read(match);
    }
  }

  return null;
}

/**
 * Collects every suit length stated in one section. The bounds are merged rather than replaced, so a description that gives a
 * floor and a ceiling for the same suit in separate breaths ends up with both.
 */
function readSuits(section: string, into: Map<SuitMarker, Bounds>): void {
  for (const { pattern, read } of suitPatterns) {
    for (const match of section.matchAll(pattern)) {
      const marker = match[2].toLowerCase() as SuitMarker;
      const current = into.get(marker) ?? { lower: null, upper: null };
      into.set(marker, { ...current, ...read(Number(match[1])) });
    }
  }
}

File: Model/Bidding/Validation/ConditionReader.cs
using Model.Enums;
using System.Text.RegularExpressions;

namespace Model.Bidding.Validation;

/// <summary>
/// Reads the suit lengths a bid promises in its own description, so the prose a human reads can be checked against the ranges
/// the engine actually acts on.
/// </summary>
/// <remarks>
/// Deliberately the same reading the Bidding Browser performs on Shift+Enter, down to the patterns, so an author who fills the
/// range fields from the text cannot then be told by the validator that the two disagree.
/// </remarks>
public static class ConditionReader {

    /// <summary>
    /// Suits are told apart by the first two letters of the word, as the descriptions write them.
    /// <para>
    /// <c>ka</c> must not swallow "kart", "kartowy" or "karty": those count cards rather than name diamonds, and they sit in
    /// exactly the position a suit would - "4+ kartowy fit" is a four card fit, not four diamonds.
    /// </para>
    /// </summary>
    private const string SuitMarkers = @"tr|ka(?!rt)|ki|pi";

    /// <summary>"5+ kierów" - a floor on the suit.</summary>
    private static readonly Regex AtLeastPattern = new($@"(\d{{1,2}})\s*\+\s*({SuitMarkers})", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    /// <summary>"brak 4 pików" - denying four means holding at most three.</summary>
    private static readonly Regex AtMostPattern = new($@"brak\s+(\d{{1,2}})\s*({SuitMarkers})", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    /// <summary>"dokładnie 5 trefli" - both ends at once.</summary>
    private static readonly Regex ExactlyPattern = new($@"dok[łl]adnie\s+(\d{{1,2}})\s*({SuitMarkers})", RegexOptions.IgnoreCase | RegexOptions.Compiled);


    /// <summary>
    /// Suit lengths stated in <paramref name="condition"/>. The text is taken a comma separated section at a time and a section
    /// that states nothing recognisable is skipped, which is what lets the prose around the numbers stay prose.
    /// </summary>
    public static IReadOnlyDictionary<BidColor, NumberRange> ReadSuitLengths(string? condition) {
        var lengths = new Dictionary<BidColor, NumberRange>();
        if (string.IsNullOrWhiteSpace(condition)) {
            return lengths;
        }

        foreach (var section in condition.Split(',')) {
            // Bounds are merged rather than replaced, so a section naming a floor and another naming a ceiling both land.
            Collect(AtLeastPattern, section, lengths, (current, count) => new NumberRange(count, current?.Upper));
            Collect(AtMostPattern, section, lengths, (current, count) => new NumberRange(current?.Lower, Math.Max(0, count - 1)));
            Collect(ExactlyPattern, section, lengths, (_, count) => new NumberRange(count, count));
        }

        return lengths;
    }


    private static void Collect(Regex pattern, string section, Dictionary<BidColor, NumberRange> lengths, Func<NumberRange?, int, NumberRange> merge) {
        foreach (Match match in pattern.Matches(section)) {
            if (!int.TryParse(match.Groups[1].Value, out var count)) {
                continue;
            }

            var suit = SuitFor(match.Groups[2].Value);
            lengths.TryGetValue(suit, out var current);
            lengths[suit] = merge(current, count);
        }
    }


    private static BidColor SuitFor(string marker) => marker.ToLowerInvariant() switch {
        "tr" => BidColor.Clubs,
        "ka" => BidColor.Diamonds,
        "ki" => BidColor.Hearts,
        _ => BidColor.Spades
    };
}
